Add traversable instance for Maybe

scott-christopher opened the pull request:
Implements #118 

As per the included docs:

**Maybe#sequence :: Applicative f => Maybe (f a) ~> (b -> f b) -> f (Maybe a)**

Evaluates an applicative action contained within the Maybe, resulting in:
  - an applicative of `Just` the value of the evaluated action if `this`
    is a `Just`.
  - a pure applicative of `Nothing` if `this` is a `Nothing`.
```javascript
> S.Nothing().sequence(S.Either.of)
Right(Nothing())

> S.Just(Right(42)).sequence(S.Either.of)
Right(Just(42))
```

davidchambers commented:
That was quick, Scott! Thanks!

davidchambers commented on index.js:
```diff
@@ -485,8 +485,8 @@
-  //. The Maybe type satisfies the [Monoid][], [Monad][], [Foldable][], and
-  //. [Extend][] specifications.
+  //. The Maybe type satisfies the [Monoid][], [Monad][], [Foldable][],
+  //. [Traversable][] and [Extend][] specifications.
```
Traversable implies Foldable, so we could remove Foldable from the list in the interest of brevity.

scott-christopher replied:
Good point – updated.

davidchambers commented on index.js:
```diff
@@ -485,7 +485,7 @@
   //. either a Just whose value is of type `a` or a Nothing (with no value).
   //.
-  //. The Maybe type satisfies the [Monoid][], [Monad][], [Foldable][], and
+  //. The Maybe type satisfies the [Monoid][], [Monad][], [Traversable][] and
```
Can we keep the Oxford comma? :)

scott-christopher replied:
:zap:

davidchambers commented on index.js:
```diff
@@ -2159,4 +2187,5 @@
 //. [Ramda]:        http://ramdajs.com/
 //. [RegExp]:       https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp
 //. [Semigroup]:    https://github.com/fantasyland/fantasy-land#semigroup
+//. [Traversable]:  https://github.com/fantasyland/fantasy-land#traversable
```
Can we remove the link definition for `[Foldable]`? I believe we're removing the only occurrence.

scott-christopher replied:
:zap:

davidchambers commented on test/index.js:
```diff
@@ -47,6 +47,45 @@ var squareRoot = function(n) {
+};
+
+//  eitherToMaybe :: Either a b -> Maybe b
+var eitherToMaybe = S.either(R.always(S.Nothing()), S.Just);
```
Let's use `S.K` in place of `R.always`.

scott-christopher replied:
:zap:

davidchambers commented on test/index.js:
```diff
@@ -579,6 +618,29 @@ describe('maybe', function() {
+    it('implements Traversable', function() {
+      var u = S.Nothing(),
+          C = Compose(Identity, S.Maybe),
+          lhs, rhs;
```
I prefer one `var` keyword per declaration, though I don't mind grouping the uninitialized variables:

```javascript
var u = S.Nothing();
var C = Compose(Identity, S.Maybe);
var lhs, rhs;
```

scott-christopher replied:
:zap:

davidchambers commented on index.js:
````diff
@@ -714,6 +714,23 @@
+  //.
+  //. > S.Just(Right(42)).sequence(S.Either.of)
+  //. Right(Just(42))
+  //. ```
````
Nice examples!

davidchambers commented on index.js:
```diff
@@ -863,6 +884,13 @@
+  //  Just#sequence :: Applicative f => Maybe (f a) ~> (b -> f b) -> f (Maybe a)
+  Just.prototype.sequence =
+    def('Just#sequence', [Function], function(of) {
+      void of;
```
`make lint` passes for me without this line, thanks to #31. Was it failing for you and/or CircleCI?

scott-christopher replied:
I've started doing this to make it clear that the argument isn't used. Happy to remove it.

scott-christopher replied:
:zap: I've removed the `void of`

davidchambers replied:
Okay. I wouldn't have minded keeping it in that case. I assumed we were appeasing some overeager linter. :)

davidchambers commented on test/index.js:
```diff
@@ -825,6 +887,30 @@ describe('maybe', function() {
+      }));
+      assert(lhs.equals(rhs));
+    });
   });
```
Please add an empty line between lines 913 and 914.

scott-christopher replied:
:zap:

davidchambers commented on index.js:
```diff
@@ -714,6 +714,23 @@
+  //.   - an applicative of `Just` the value of the evaluated action if `this`
+  //.     is a `Just`.
+  //.
+  //.   - a pure applicative of `Nothing` if `this` is a `Nothing`.
```
It would be nice to mirror the structure of the description of `Maybe#reduce`:

> Evaluates an applicative action contained within the Maybe, resulting in:
>
>   - a pure applicative of a Nothing if `this` is a Nothing; otherwise
>
>   - an applicative of Just the value of the evaluated action.

scott-christopher replied:
:zap:

davidchambers commented:
Lovely pull request, Scott. Now that I've appeased my pedantic side it's time to spend some time studying these laws to make sure I understand them. :)

scott-christopher commented:
This loose translation from [here](https://hackage.haskell.org/package/base-4.8.1.0/docs/Data-Traversable.html#t:Traversable) _might_ help with interpreting the laws:

---

A definition of `sequence` must satisfy the following laws:

**naturality**
* `compose(t, sequence) = compose(sequence, map(t))` for every applicative transformation `t`, where an applicative transformation is a function: `t :: (Applicative f, Applicative g) => f a -> g a`

**identity**
* `compose(sequence, map(Identity)) = Identity`

**composition**
* `compose(sequence, map(Compose)) = compose(Compose, map(sequence), sequence)`

davidchambers commented on test/index.js:
```diff
@@ -47,6 +47,45 @@ var squareRoot = function(n) {
+var Compose = function(F, G) {
+  function _Compose(x) {
+    return {
+      map: function(fn) {
```
I've a slight preference for `f` over `fn`. Mind changing this, Scott?

scott-christopher replied:
:zap:

davidchambers commented on test/index.js:
```diff
@@ -579,6 +618,29 @@ describe('maybe', function() {
 
+    it('implements Traversable', function() {
+      var u = S.Nothing();
+      var C = Compose(Identity, S.Maybe);
```
Should this be `S.Maybe.of` rather than `S.Maybe`?

scott-christopher replied:
You can think of the arguments to `Compose` as the types that are expected to have an `of` method attached which are used to implement `of` for `Compose`, so we want to pass `Identity` and `S.Maybe` here.

```js
_Compose.of = function(x) {
  return _Compose(F.of(G.of(x)));
};
```

davidchambers replied:
Oh I see. Thanks for the explanation. :)

davidchambers commented:
I'm sorry for neglecting this pull request, Scott! Would you mind rebasing?

scott-christopher commented:
> I'm sorry for neglecting this pull request

I forgot about it myself :)

I've rebased on to master and also updated the tests for `sequence` to use jsverify.

scott-christopher commented on index.js:
```diff
@@ -831,6 +831,29 @@
+  Maybe.prototype.sequence =
+  method('Maybe#sequence',
+         {a: [Apply], b: [Apply]},
+         [$Maybe(a), $.Function, b],
```
Is there a better way to represent `Apply f => Maybe (f a) -> (a -> f a) -> f (Maybe a)` here?

davidchambers replied:
I don't think so. We're quite a bit beyond the city limits of sanctuary-def at this point. ;)

davidchambers commented:
>       1) maybe Maybe implements Traversable:
>          Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test.

Looks as though we need a bit more :hourglass:.

scott-christopher commented:
I've split them out into separate `it` blocks to give them each a little more time to run.

davidchambers commented on index.js:
```diff
@@ -540,8 +540,8 @@
-  //. The Maybe type satisfies the [Monoid][], [Monad][], [Foldable][], and
-  //. [Extend][] specifications.
+  //. The Maybe type satisfies the [Monoid][], [Monad][], [Foldable][],
+  //. [Traversable][], and [Extend][] specifications.
```
This came back to life during the rebase:

> Traversable implies Foldable, so we could remove Foldable from the list in the interest of brevity.

I believe this comment still applies:

> Can we remove the link definition for `[Foldable]`? I believe we're removing the only occurrence.

scott-christopher replied:
Haha, I looked at this yesterday and couldn't remember why `Foldable` was removed previously and presumed `reduce` hadn't yet landed at that time. That'll teach me for not looking at the comment history :smile: 

In any case, I've removed it again.

davidchambers commented on test/index.js:
```diff
@@ -67,6 +67,69 @@ var squareRoot = function(n) {
                : S.Right(Math.sqrt(n));
 };
 
+var Identity = function(x) {
```
A type signature here would be nice:

```javascript
//  Identity :: a -> Identity a
var Identity = ...;
```

scott-christopher replied:
Sure. :zap: I've included signatures for `Compose` and the Arbitrary constructors too.

davidchambers commented on test/index.js:
```diff
@@ -478,6 +541,37 @@ describe('maybe', function() {
+      });
+
+    });
+
```
I :heart: these tests!

davidchambers commented on index.js:
```diff
@@ -831,6 +831,29 @@
+         {a: [Apply], b: [Apply]},
+         [$Maybe(a), $.Function, b],
+         function(maybe, of) {
+           return maybe.isJust ? maybe.value.map(Just) : of(maybe);
```
I think we should use `R.map(Just, maybe.value)` here, to support environments without `Array.prototype.map`.

scott-christopher replied:
Good idea. :zap:

davidchambers commented on test/index.js:
```diff
@@ -67,6 +67,79 @@ var squareRoot = function(n) {
+
+//  Compose :: Apply f, Apply g
+//          => { of: b -> f b } -> { of: c -> g c }
+//          -> f (g a) -> Compose f g a
```
This is instructive!

davidchambers commented:
:deciduous_tree:
